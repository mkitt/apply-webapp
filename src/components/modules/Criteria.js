// @flow
import React, { type Node } from 'react'
import glamorous from 'glamorous'
import type { Location } from 'react-router-dom'
import { map } from 'ramda'
import CriterionModule from './Criterion'
import Loader from '../blocks/Loader'
import SectionContainment from '../elements/SectionContainment'
import type { Criterion } from '../../types/Criterion'

type Props = {
  children: Node,
  criteria: Array<Criterion>,
  handleChange: () => {},
  location: Location,
  refetch: () => {},
}

export default class extends React.PureComponent<Props> {
  componentWillReceiveProps(nextProps: Props) {
    if (this.props.location.key !== nextProps.location.key) {
      this.props.refetch()
    }
  }

  render() {
    const { criteria, children, handleChange } = this.props
    return (
      <SectionContainment>
        <Criteria>
          {map(criterion => (
            <CriterionModule criterion={criterion} handleChange={handleChange} key={criterion.id} />
          ), criteria)}
        </Criteria>
        {children}
      </SectionContainment>
    )
  }
}

export const ModuleLoader = () => (
  <Loader>Loading Evaluations...</Loader>
)

// -------------------------------------

const Criteria = glamorous.section({
  display: 'flex',
  flexDirection: 'column',
  paddingTop: 6,
  paddingBottom: 32,
  '& > *': {
    marginTop: 48,
  },
})
