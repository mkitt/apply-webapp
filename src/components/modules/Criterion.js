// @flow
import React from 'react'
import glamorous from 'glamorous'
import { memoizeWith, identity, length } from 'ramda'
import { pure } from 'recompose'
import { COLORS } from '../../constants/theme'
import { ChevronXIcon } from '../elements/Icons'
import type { Criterion } from '../../types/Criterion'

type Props = {
  criterion: Criterion,
  isOpen: boolean,
}

const len = memoizeWith(identity, list => length(list) - 1)

export default pure(({ criterion, isOpen }: Props) => (
  <Details open={isOpen}>
    <Summary>
      <span>{criterion.label}</span>
      <ChevronXIcon />
    </Summary>
    <Content>
      <Questions>
        {criterion.questions.map(question => (
          <Question key={question}>
            {question}
          </Question>
        ))}
      </Questions>
      <OptionDl>
        {criterion.options.map((option, index) => (
          <OptionItem key={option}>
            <OptionDt>{len(criterion.options) - index}</OptionDt>
            <OptionDd>{option}</OptionDd>
          </OptionItem>
        ))}
      </OptionDl>
    </Content>
  </Details>
))

// -------------------------------------

const Details = glamorous.details({
})

const Summary = pure(glamorous.summary({
  position: 'relative',
  minHeight: 48,
  fontSize: 28,
  fontWeight: 200,
  color: COLORS.GREY_8,
  borderBottom: '1px solid',
  outline: 0,
  cursor: 'pointer',
  '&::-webkit-details-marker': {
    display: 'none',
  },
  '&:first-of-type': {
    listStyleType: 'none',
  },
  'details[open] > &': {
    color: COLORS.GREY_4,
  },
  '& > svg': {
    position: 'absolute',
    top: 8,
    right: 0,
  },
  // Yo! Using a `not` filter here
  'details:not([open]) > & > svg': {
    top: 16,
  },
  // Yo! Also using a `not` filter here
  'details:not([open]) > & > svg g:nth-child(2)': {
    opacity: 0,
  },
}))

const Content = glamorous.div({
  paddingTop: 44,
  paddingBottom: 48,
})

const Questions = glamorous.ol({
  listStyle: 'none',
  margin: 0,
  paddingTop: 0,
  paddingRight: 0,
  paddingBottom: 38,
  paddingLeft: 48,
  color: COLORS.GREY_4,
})

const Question = glamorous.li({
  fontSize: 16,
  margin: 0,
  marginBottom: 24,
})

const OptionDl = glamorous.dl({
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'flex-start',
  minHeight: 160,
  margin: 0,
  marginTop: 14,
  paddingTop: 32,
  paddingRight: 48,
  paddingBottom: 42,
  paddingLeft: 48,
  color: COLORS.GREY_4,
  backgroundColor: COLORS.GREY_E,
})

const OptionItem = glamorous.div({
  paddingRight: 32,
})

const OptionDt = glamorous.dt({
  marginBottom: 14,
  fontSize: 28,
  fontWeight: 200,
})

const OptionDd = glamorous.dd({
  marginLeft: 0,
})
