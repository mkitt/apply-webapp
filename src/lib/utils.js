// @flow
import { __, addIndex, always, and, complement, either, gt, isEmpty, isNil, map } from 'ramda'
import { withProps } from 'recompose'

// -------------------------------------
// Environments
export const isDev = always(process.env.NODE_ENV === 'development')

// -------------------------------------
// Types
export const isNotNil = complement(isNil)
export const isNotEmpty = complement(isEmpty)
export const isEmptyOrNil = either(isNil, isEmpty)
export const isNotEmptyAndNotNil = and(isNotNil, isNotEmpty)

// -------------------------------------
// Relation
export const gt0 = gt(__, 0)

// -------------------------------------
// Lists
export const nool = always([])()
export const mapIndexed = addIndex(map)

// -------------------------------------
// Objects
export const noob = always({})()

// -------------------------------------
// Functions
export const noop = () => {}

export const log = withProps(props => (
  // eslint-disable-next-line no-console
  console.log(props.displayName || 'üëç', props) && props
))

export const warn = withProps(props => (
  // eslint-disable-next-line no-console
  console.warn(props.displayName || '‚ö†Ô∏è', props) && props
))
