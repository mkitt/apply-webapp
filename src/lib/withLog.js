import { withProps } from 'recompose'

export default withProps(props => (
  // eslint-disable-next-line
  console.log(props.logName || props.displayName || '👍', props) && props
))
