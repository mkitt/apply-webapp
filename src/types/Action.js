// @flow
import type { Applicant } from './Applicant'

export type Action = {
  applicants: Array<Applicant>,
  id: ID,
  label: string,
  message: string,
  name: string,
}
