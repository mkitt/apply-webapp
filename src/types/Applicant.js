// @flow
import { camelize } from 'humps'
import { compose, prop, split, head, last } from 'ramda'
import type { Action } from './Action'
import type { Application } from './Application'

export default undefined

export type Applicant = {
  action: ?Action,
  applications: Array<Application>,
  birthDate: ?string,
  createdAt: DateTime,
  email: string,
  github: string,
  id: ID,
  name: string,
  referredBy: ?string,
  resume: ?string,
  startDate: ?DateTime,
  status: ?string,
  updatedAt: DateTime,
  // derived data...
  firstName: string,
  lastName: string,
}

// -------------------------------------
// Derived helpers

export const deriveFirstName = compose(
  head,
  split(' '),
  prop('name'),
)

export const deriveLastName = compose(
  last,
  split(' '),
  prop('name'),
)

// -------------------------------------
// For testing content
export const stub = (props: Applicant) => {
  const firstName = (props != null && props.firstName) || 'Kenny'
  const lastName = (props != null && props.lastName) || 'Bania'
  const camel = camelize(`${firstName} ${lastName}`)
  const applicant = {
    id: camel,
    action: null,
    applications: [],
    birthDate: '06/06/2006',
    createdAt: '2010-10-31T12:12:12.000Z',
    email: 'kenny.bania@example.com',
    firstName: 'Kenny',
    github: camel,
    lastName,
    referredBy: 'Simon Taranto',
    resume: 'https://www.resume.com',
    startDate: null,
    status: 'pending',
    updatedAt: '2010-12-24T24:24:24.000Z',
  }
  return {
    ...applicant,
    ...props,
  }
}
