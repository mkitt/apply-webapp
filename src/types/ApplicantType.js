// @flow
export type ApplicantType = {
  applyAction: string,
  applyStatus: string,
  birthdate: string,
  createdAt: string,
  email: string,
  firstName: string,
  github: string,
  id: string,
  lastName: string,
  referredBy: string,
  resume: string,
  scoreLogicEvaluation: number,
  scoreOnlineLogicTest: number,
  scoreValuesEvaluation: number,
  startDate: string,
  updatedAt: string,
}
