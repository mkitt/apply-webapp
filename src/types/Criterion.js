// @flow
import { camelize } from 'humps'
import { compose, prop } from 'ramda'
import type { Evaluation } from './Evaluation'

export type Criterion = {
  id: ID,
  createdAt: DateTime,
  evaluation: ?Evaluation,
  title: string,
  options: Array<string>,
  questions: Array<string>,
  updatedAt: DateTime,
  // derived...
  name: string,
}

// -------------------------------------
// Derived helpers

export const deriveName = compose(
  camelize,
  prop('title'),
)

// -------------------------------------
// For testing content
export const stub = (props: Criterion) => {
  const title = (props != null && props.title) || 'Problem Solving'
  const camel = camelize(title)
  return {
    id: camel,
    title,
    name: camel,
    options: [
      `${title} Option 3`,
      `${title} Option 2`,
      `${title} Option 1`,
    ],
    questions: [
      `${title} Question 1`,
      `${title} Question 2`,
      `${title} Question 3`,
    ],
    ...props,
  }
}

export default undefined
