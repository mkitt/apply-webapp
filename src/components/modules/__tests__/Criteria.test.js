import React from 'react'
import renderer from 'react-test-renderer'
import Criteria from '../Criteria'
import { stub } from '../../../types/Criterion'
import { noop } from '../../../lib/utils'

const criteria = [
  stub({ label: 'Problem Solving' }),
  stub({ label: 'Self Reflection' }),
]

describe('<Criteria />', () => {
  it('shows the correct snapshot tree for the <Criteria> module', () => {
    const component = (
      <Criteria
        criteria={criteria}
        handleChange={{ noop }}
        location={{ key: '1' }}
        refetch={{ noop }}
      />
    )
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
