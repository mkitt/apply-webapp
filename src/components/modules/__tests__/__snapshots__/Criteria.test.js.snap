// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Criteria /> shows the correct snapshot tree for the <Criteria> module 1`] = `
.css-0,
[data-css-0] {
  position: relative;
  max-width: 960px;
  margin-right: auto;
  margin-left: auto;
  padding-right: 16px;
  padding-left: 16px;
}

.css-1,
[data-css-1] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  flex-direction: column;
  padding-top: 6px;
  padding-bottom: 32px;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -webkit-flex-direction: column;
}

.css-1 > *,
[data-css-1] > * {
  margin-top: 48px;
}

.css-2,
[data-css-2] {
  position: relative;
  min-height: 48px;
  font-size: 28px;
  font-weight: 200;
  color: #808080;
  border-bottom: 1px solid;
  outline: 0;
  cursor: pointer;
}

.css-2::-webkit-details-marker,
[data-css-2]::-webkit-details-marker {
  display: none;
}

.css-2:first-of-type,
[data-css-2]:first-of-type {
  list-style-type: none;
}

.css-2 > svg,
[data-css-2] > svg {
  position: absolute;
  top: 8px;
  right: 0;
}

.css-5,
[data-css-5] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: space-between;
  padding-right: 48px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
}

.css-5 > span + span,
[data-css-5] > span + span {
  color: #35c5fc;
}

.css-6,
[data-css-6] {
  position: relative;
  padding-top: 44px;
  padding-bottom: 48px;
}

.css-7,
[data-css-7] {
  width: 256px;
}

.css-8,
[data-css-8] {
  position: relative;
}

.css-9,
[data-css-9] {
  -webkit-appearance: button;
  -webkit-border-radius: 0;
  -webkit-box-shadow: none;
  -webkit-padding-end: 0;
  -webkit-padding-start: 0;
  -webkit-user-select: none;
  width: 100%;
  height: 48px;
  margin: 0;
  padding: 0 32px;
  overflow: hidden;
  font-size: 14px;
  font-family: inherit;
  text-transform: uppercase;
  color: #fff;
  background-color: #4d4d4d;
  background-image: none;
  background-position: initial;
  background-repeat: no-repeat;
  border-radius: 0;
  border: 0;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.css-9:focus,
[data-css-9]:focus {
  border-color: #999;
  outline: 0;
}

.css-11,
[data-css-11] {
  position: absolute;
  top: calc(50% - 8px);
  right: 32px;
  color: #fff;
  pointer-events: none;
}

.css-12,
[data-css-12] {
  list-style: none;
  margin: 0;
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 38px;
  padding-left: 48px;
  color: #4d4d4d;
}

.css-13,
[data-css-13] {
  font-size: 16px;
  margin: 0;
  margin-bottom: 24px;
}

.css-14,
[data-css-14] {
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;
  justify-content: space-between;
  align-items: flex-start;
  min-height: 160px;
  margin: 0;
  margin-top: 14px;
  padding-top: 32px;
  padding-right: 48px;
  padding-bottom: 42px;
  padding-left: 48px;
  color: #4d4d4d;
  background-color: #e5e5e5;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -webkit-box-align: start;
  -webkit-align-items: flex-start;
}

.css-15,
[data-css-15] {
  padding-right: 32px;
}

.css-16,
[data-css-16] {
  margin-bottom: 14px;
  font-size: 28px;
  font-weight: 200;
}

.css-17,
[data-css-17] {
  margin-left: 0;
}

@media (min-width: 48em) {
  .css-7,
  [data-css-7] {
    position: absolute;
    right: 0;
  }
}

<section
  className="css-0"
>
  <section
    className="css-1"
  >
    <details
      className="css-nil"
      open={undefined}
    >
      <summary
        className="css-2"
      >
        <div
          className="css-5"
        >
          <span>
            Problem Solving
          </span>
          <span>
            
          </span>
        </div>
        <svg
          className="css-nil"
          height={28}
          viewBox="0 0 28 28"
          width={28}
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
          >
            <polyline
              points="27.155 0 14.224 12.931 0 0"
            />
            <g
              transform="translate(13.577500, 19.465500) scale(1, -1) translate(-13.577500, -19.465500)"
            >
              <polyline
                points="27.155 13 14.224 25.931 0 13"
              />
            </g>
          </g>
        </svg>
      </summary>
      <div
        className="css-6"
      >
        <div
          className="css-7"
        >
          <label
            className="css-8"
            htmlFor="problemSolving"
          >
            <select
              aria-label="Score"
              className="css-9"
              id="problemSolving"
              name="problemSolving"
              onChange={
                Object {
                  "noop": [Function],
                }
              }
              value={undefined}
            >
              <option
                value={null}
              >
                Score
              </option>
              <option
                value={0}
              >
                0 - Bad
              </option>
              <option
                value={1}
              >
                1 - Meh
              </option>
              <option
                value={2}
              >
                2 - Good
              </option>
              <option
                value={3}
              >
                3 - Excellent
              </option>
            </select>
            <div
              className="css-11"
            >
              <svg
                className="css-nil"
                height={12}
                viewBox="0 0 21 12"
                width={21}
              >
                <g
                  fill="none"
                  fillRule="evenodd"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                >
                  <polyline
                    points="19.182 1.067 10.611 10.067 1.182 1.067"
                  />
                </g>
              </svg>
            </div>
          </label>
        </div>
        <ol
          className="css-12"
        >
          <li
            className="css-13"
          >
            Problem Solving Question 1
          </li>
          <li
            className="css-13"
          >
            Problem Solving Question 2
          </li>
          <li
            className="css-13"
          >
            Problem Solving Question 3
          </li>
        </ol>
        <dl
          className="css-14"
        >
          <div
            className="css-15"
          >
            <dt
              className="css-16"
            >
              2
            </dt>
            <dd
              className="css-17"
            >
              Problem Solving Option 3
            </dd>
          </div>
          <div
            className="css-15"
          >
            <dt
              className="css-16"
            >
              1
            </dt>
            <dd
              className="css-17"
            >
              Problem Solving Option 2
            </dd>
          </div>
          <div
            className="css-15"
          >
            <dt
              className="css-16"
            >
              0
            </dt>
            <dd
              className="css-17"
            >
              Problem Solving Option 1
            </dd>
          </div>
        </dl>
      </div>
    </details>
    <details
      className="css-nil"
      open={undefined}
    >
      <summary
        className="css-2"
      >
        <div
          className="css-5"
        >
          <span>
            Self Reflection
          </span>
          <span>
            
          </span>
        </div>
        <svg
          className="css-nil"
          height={28}
          viewBox="0 0 28 28"
          width={28}
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
          >
            <polyline
              points="27.155 0 14.224 12.931 0 0"
            />
            <g
              transform="translate(13.577500, 19.465500) scale(1, -1) translate(-13.577500, -19.465500)"
            >
              <polyline
                points="27.155 13 14.224 25.931 0 13"
              />
            </g>
          </g>
        </svg>
      </summary>
      <div
        className="css-6"
      >
        <div
          className="css-7"
        >
          <label
            className="css-8"
            htmlFor="selfReflection"
          >
            <select
              aria-label="Score"
              className="css-9"
              id="selfReflection"
              name="selfReflection"
              onChange={
                Object {
                  "noop": [Function],
                }
              }
              value={undefined}
            >
              <option
                value={null}
              >
                Score
              </option>
              <option
                value={0}
              >
                0 - Bad
              </option>
              <option
                value={1}
              >
                1 - Meh
              </option>
              <option
                value={2}
              >
                2 - Good
              </option>
              <option
                value={3}
              >
                3 - Excellent
              </option>
            </select>
            <div
              className="css-11"
            >
              <svg
                className="css-nil"
                height={12}
                viewBox="0 0 21 12"
                width={21}
              >
                <g
                  fill="none"
                  fillRule="evenodd"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                >
                  <polyline
                    points="19.182 1.067 10.611 10.067 1.182 1.067"
                  />
                </g>
              </svg>
            </div>
          </label>
        </div>
        <ol
          className="css-12"
        >
          <li
            className="css-13"
          >
            Self Reflection Question 1
          </li>
          <li
            className="css-13"
          >
            Self Reflection Question 2
          </li>
          <li
            className="css-13"
          >
            Self Reflection Question 3
          </li>
        </ol>
        <dl
          className="css-14"
        >
          <div
            className="css-15"
          >
            <dt
              className="css-16"
            >
              2
            </dt>
            <dd
              className="css-17"
            >
              Self Reflection Option 3
            </dd>
          </div>
          <div
            className="css-15"
          >
            <dt
              className="css-16"
            >
              1
            </dt>
            <dd
              className="css-17"
            >
              Self Reflection Option 2
            </dd>
          </div>
          <div
            className="css-15"
          >
            <dt
              className="css-16"
            >
              0
            </dt>
            <dd
              className="css-17"
            >
              Self Reflection Option 1
            </dd>
          </div>
        </dl>
      </div>
    </details>
  </section>
</section>
`;
